### ISSUE TITLE
Feature request: smarter primitive for container_types

### ISSUE THREAD

**Issue by jackkamm** [HIPS/autograd#99]
I've been noticing that `grad()` can be quite memory intensive, due to the tape.

As a workaround, I've been making some of my subroutines `@primitive`, and assigning their derivatives thru `defgrad`. This way, the gradient for the subroutine gets computed "off" the tape.

One issue is if the primitive subroutine takes a dict containing nodes, i.e. `subroutine(a_dict)`. Then `primitive.__call__()` doesn't know to call `subroutine.gradmaker` because `not isinstance(a_dict, Node)`. The result is the computations in `subroutine` continue to be added to the tape.

My solution is to convert `a_dict` into a `DictNode`; then the subroutine truly acts as a primitive.

It would be nice if `primitive.__call__()` did this automatically (i.e., not just check if the arguments are Nodes, but if they are container types with Nodes inside). Otherwise, primitives that take container_types aren't really primitive.

**Comment by jackkamm** [comment]
Hmm, I just realized I could make my primitive `subroutine` have signature
```
@primitive
def subroutine(keys, *vals):
    a_dict = dict(zip(keys, vals))
    ...
```
instead of
```
@primitive
def subroutine(a_dict):
    ...
```
which would also solve this issue.

It might still be neat to have `primitive` work with container_type arguments but maybe it would be too much work/code bloat. I'll leave it to you to decide whether to close this issue.

**Comment by mattjj** [comment]
I agree with your diagnosis: in general, there are cases where we'd need to recursively descend into data structures to look for any nodes in them. I've run into some corner cases along these lines, especially when making tuples of nodes and when defining my own primitives; to ensure that they get wrapped into a TupleNode, I call `make_tuple`, which I put in `autograd.util`. I think your `make_dict` is similar, and we should probably add that to `autograd.util` too. (We've also seen similar nested data structure challenges come up when primitive arguments are bundled into `kwargs` in such a way that autograd doesn't unbox nodes before passing them to the wrapped function, like with some of the scipy wrappers.)

However, recursively descending into data structures would make the boxing/unboxing logic a lot more complicated (not just the implementation but even the decision of what cases to handle) and also slower for the common case. As long as these issues remain rare corner cases, requiring the power users to call `make_tuple` or `make_dict` seems pretty reasonable, at least until we find a better solution.

Dougal's been cleaning up the core logic a bit in a branch. It may be the changes that he's made will help this container type logic. For the time being, I'm going to add `make_dict` to `autograd.util` and close this issue.

**Comment by mattjj** [comment]
For posterity, [this commit](https://github.com/HIPS/autograd/commit/6f1d8f2b13a77185df8f4b10fc2e2e0a0e6a6309) is where we added `make_tuple`.

**Comment by mattjj** [comment]
On second thought, I'm not going to add `make_dict` yet; let's see how the new core logic might treat things differently.

**Comment by jackkamm** [comment]
Sounds reasonable -- doing all the extra checking could certainly slow down a lot of common cases. I think requiring power users to call `make_tuple` or `make_dict` to move computations off the tape is pretty reasonable.

Maybe this functionality for moving certain computations off the tape, could become an official part of the API/documentation/functionality, with some unit tests? In practice it can be hard to tell whether a computation is happening on or off the tape, so my worry is that future changes to autograd may affect the code behavior in a way that is hard to diagnose.

Anyways, thanks for the response, I was worried my initial rambling post would be impossible to understand :)

**Comment by mattjj** [comment]
We read you loud and clear; we've just gotten busy with thesis defenses and job hunts.

I think "moving computations off the tape" means defining your own primitives and, in so doing, being able to discard some values that aren't needed on the reverse pass. Defining your own primitives is in the documentation (to the extent that anything here is documented), but there are some more things we could write down and some API adjustments we could make. In particular, sometimes your primitive does need to cache some results for the reverse pass gradfun to use, and the API should support that. I hacked something like that into the API as [`primitive_with_aux`](https://github.com/HIPS/autograd/blob/42093411684f75d941805c6e4eba5702311a7706/autograd/core.py#L150-188) and used it for [a project where I wrote some primitives and gradients in cython](https://github.com/mattjj/svae), but that implementation isn't a good one because it breaks higher-order differentiation. Now that we understand this issue better we're hoping that an updated core design will handle these cases.
